Unique Exception

fogelito opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a distinct Unique error type to represent non-primary unique-index violations.

* **Bug Fixes**
  * Adapters now distinguish primary-identifier duplicates from other unique-index conflicts; non-primary conflicts surface as Unique errors.
  * Duplicate/error handling consolidated into a common processor for more consistent and clearer error reporting across databases.
  * Public API signatures unchanged.

* **Tests**
  * End-to-end tests updated to accept and assert both Duplicate and Unique error outcomes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

fogelito committed:
UniqueException

fogelito committed:
SQLite.php Exception

fogelito committed:
formatting

fogelito committed:
SQLite tests

fogelito committed:
SQLite tests

fogelito committed:
Mysql

fogelito committed:
Mysql

fogelito committed:
SQLite.php

fogelito committed:
SQLite.php

fogelito committed:
Remove var_dump

fogelito committed:
Postgres.php

fogelito committed:
Mysql Use str_ends_with

fogelito committed:
Fix mariaDB str_ends_with

fogelito committed:
Merge branch 'main' of github.com:utopia-php/database into unique-exception

# Conflicts:
#	tests/e2e/Adapter/Scopes/DocumentTests.php

Copilot committed:
Fix UniqueException message to say 'Unique index violation' instead of 'Document already exists'

Agent-Logs-Url: https://github.com/utopia-php/database/sessions/bd2e3ddb-ca02-49dc-9939-bb42093478e2

Co-authored-by: abnegate <5857008+abnegate@users.noreply.github.com>

abnegate committed:
Merge pull request #856 from utopia-php/copilot/fix-unique-exception-message

Fix UniqueException message to indicate unique index violation

abnegate commented on src/Database/Exception/Unique.php:
```diff
@@ -0,0 +1,7 @@
+
+namespace Utopia\Database\Exception;
+
+class Unique extends Duplicate
```
@copilot Change this to extend the root database exception instead of duplicate

abnegate replied:
Keeping `Unique extends Duplicate` deliberately: every existing call site that catches `Duplicate` on `createDocument`/`updateDocument` keeps working, and callers that need the distinction narrow with `instanceof Unique`. Extending the root `Exception` instead would silently turn all current unique-index catches into uncaught exceptions across Appwrite.

Copilot committed:
Address review feedback: Unique extends Exception, MariaDB uses str_contains for _uid check

Agent-Logs-Url: https://github.com/utopia-php/database/sessions/20af0e15-866e-40a9-b3da-a67b9c498fa7

Co-authored-by: abnegate <5857008+abnegate@users.noreply.github.com>

abnegate commented on src/Database/Adapter/SQLite.php:
```diff
@@ -1248,9 +1241,21 @@ protected function processException(PDOException $e): \Exception
+
+                if ($columns === ['_tenant', '_uid'] || in_array('_uid', $columns)) {
+                    return new DuplicateException('Document already exists', $e->getCode(), $e);
+                }
```
Fixed in 23eaafb6. SQLite now parses the failed column list out of `UNIQUE constraint failed: tbl.a, tbl.b`, strips the table prefix, sorts, and compares the exact set against `["_uid"]` / `["_tenant", "_uid"]`. A composite user index containing `$id` (e.g. `["$id", "email"]`) no longer matches and correctly yields `Unique`. Regression coverage: `testSQLiteCompositeIndexOnDocumentIdIsUnique` in tests/unit/UniqueViolationTest.php.

abnegate commented on src/Database/Adapter/Postgres.php:
```diff
@@ -1916,7 +1917,16 @@ protected function processException(PDOException $e): \Exception
+                $columns = array_map('trim', explode(',', $matches[1]));
+                sort($columns);
+                $target = $this->sharedTables ? ['_tenant', '_uid'] : ['_uid'];
+                if ($columns == $target) {
+                    return new DuplicateException('Document already exists', $e->getCode(), $e);
+                }
+            }
+
+            return new UniqueException('Unique index violation', $e->getCode(), $e);
         }
```
Fixed in 23eaafb6. `getViolatedColumns()` returns `null` when the `DETAIL: Key (...)` line is absent or unparsable, and `processException` falls back to `DuplicateException` in that case — pre-PR behaviour is preserved, no silent downgrade. Regression coverage: `testPostgresMissingDetailIsDuplicate` in tests/unit/UniqueViolationTest.php.

abnegate commented on src/Database/Adapter/Postgres.php:
```diff
@@ -2161,7 +2162,7 @@ protected function processException(PDOException $e): \Exception
             if (!\str_contains($message, '_uid')) {
-                return new DuplicateException('Document with the requested unique attributes already exists', $e->getCode(), $e);
+                return new UniqueException('Unique index violation', $e->getCode(), $e);
             }
```
Fixed in 23eaafb6. Postgres now parses the `DETAIL: Key (cols)=(vals)` column list and compares the sorted set against `["_uid"]` / `["_tenant", "_uid"]` rather than substring-matching the whole message, so neither a constraint name nor a conflicting value containing `_uid` can be mistaken for a document-ID conflict. Same treatment applied to MariaDB/MySQL (key name), SQLite (failed column list) and Mongo (index name). Regression coverage in tests/unit/UniqueViolationTest.php.
